Gitpod public API docs (#2939)
gitpod-io/website: gitpod/docs/references/gitpod-public-api.md

---
section: gitpod-public-api
title: Gitpod public API
---

<script context="module">
  export const prerender = true;
</script>

# Gitpod public API

> > **⚠️ Warning:** Public API is currently in development and may change frequently before it reaches General Availability.

Reviewer: Some small wording suggestions. 

```suggestion
> **⚠️ Warning:** The Public API is currently in active development. Gitpod may change the API contract **at any time** before the API is finally released as Generally Available. Use the API at your own risk. Also see [Release Cycle](/docs/help/public-roadmap/release-cycle) for the guarantees that Gitpod make for our features at various points in our release cycle. 
```

## About the Gitpod API

> It helps you to manage Gitpod Workspaces directly through gRPC APIs. Currently, all of the Authorization is on the server side, but it would be changed in future & will require a `authToken` (_coming soon_) to make an request to the APIs.

Reviewer: Can we mention the current state of auth and client libraries? Doesn't matter if things are not complete, if they're not complete, let's mention that. This documentation should be a representation of the current truth, no worries if the state is not fully complete, as this first doc is only a draft to get visibility on the current state 🙏 

```suggestion
It helps you to manage Gitpod Workspaces directly through gRPC APIs. Currently, all of the Authorization is on the server side, but it would be changed in future & will require a `authToken` (_coming soon_) to make an request to the APIs.

### Auth

Currently, auth is...

### Client Libraries

Currently, we support... 

We are looking to support...

```

Author: Resolved in [**`29fece8`**](https://github.com/gitpod-io/website/pull/2939/commits/29fece84b79c20fb8273cdcd79fa59cc5eadfeec)

| `workspaceId` |                                                       Workspace Id                                                       | string |
|   `ownerId`   |                                                         User Id                                                          | string |
|  `projectId`  |                                                        Project Id                                                        | string |
|   `context`   |                                 `contextUrl` with details of git or prebuild or snapshot                                 | string |
|   `status`    | will return Instance Id if currently assigned Instance Id to this workspace. Empty when there is no Instance Id assigned | string |

### Create & start workspace

It creates and start a new/ old workspace to the authenticated user.

**Usage**: `createAndStartWorkspace()`

**Request Parameters**:

|     Parameter      |                                Description                                 |  Type  | Required |
| :----------------: | :------------------------------------------------------------------------: | :----: | :------: |

> | `idempotencyToken` | Token that is used by the server to identify and discard replayed requests | string |   true   |

Reviewer 2: Is this `idempotencyToken` just some random string the caller provides?

Author: yeah, its a UUID used on the server side.

---
section: gitpod-public-api
title: Gitpod public API
---

<script context="module">
  export const prerender = true;
</script>

# Gitpod public API

> **⚠️ Warning:** The Public API is currently in active development. Gitpod may change the API contract **at any time** before the API is finally released as Generally Available. Use the API at your own risk. Also see [Release Cycle](/docs/help/public-roadmap/release-cycle) for the guarantees that Gitpod make for our features at various points in our release cycle.

## About the Gitpod API

> It helps you to manage Gitpod Workspaces directly through gRPC APIs. Currently, all of the Authorization is on the server side, but it would be changed in future & will require a `authToken` (_coming soon_) to make an request to the APIs.

Reviewer: ```suggestion
Gitpod public API automating the management of Gitpod Workspaces directly via gRPC APIs. 
```

## Authentication

> Currently, Authentication is under development and will be coming soon.

Reviewer: ```suggestion
Authentication is currently under development. 
```

## Client libraries support

> Currently, We only support JavaScript/TypeScript (Node). But, we are looking to support for Go.

Reviewer: ```suggestion
Public API currently supports JavaScript/TypeScript (Node). Client libraries in Go are currently planned. 
```

| :-----------: | :----------: | :----: | :------: |
| `workspaceId` | Workspace Id | string |   true   |

<br>

**Response Parameters**:

|   Parameter   |                                                       Description                                                        |  Type  |
| :-----------: | :----------------------------------------------------------------------------------------------------------------------: | :----: |
| `workspaceId` |                                                       Workspace Id                                                       | string |
|   `ownerId`   |                                                         User Id                                                          | string |
|  `projectId`  |                                                        Project Id                                                        | string |
|   `context`   |                                 `contextUrl` with details of git or prebuild or snapshot                                 | string |
|   `status`    | will return Instance Id if currently assigned Instance Id to this workspace. Empty when there is no Instance Id assigned | string |

> ### Create & start workspace

Reviewer: **Thought/question:** Any reason to merge "create" and "start" nomenclature here? Why not just "workspace start"? Or why not "workspaceStatus" with the property as "start", etc? 

See [related PR on workspace loading process](https://github.com/gitpod-io/website/pull/2858), and [Netlify Preview](https://deploy-preview-2858--gitpod-kumquat.netlify.app/docs/introduction/learn-gitpod/workspace-lifecycle)

> ---

Reviewer: Notes from discussion with @easyCZ:

* Parameters are not always camel case (depending on the language you use)
* We might be able to code generate definitions (so we don't have to hand craft the API definitions)
* Current endpoints are not entirely designed
* Start, stop (create and start) have been removed from the API

section: gitpod-public-api
title: Gitpod public API
---

<script context="module">
  export const prerender = true;
</script>

# Gitpod public API

> **⚠️ Warning:** The Public API is currently in active development. Gitpod may change the API contract **at any time** before the API is finally released as Generally Available. Use the API at your own risk. Also see [Release Cycle](/docs/help/public-roadmap/release-cycle) for the guarantees that Gitpod make for our features at various points in our release cycle.

## About the Gitpod API

It helps you to manage Gitpod Workspaces directly through gRPC APIs. Currently, all of the Authorization is on the server side, but it would be changed in future & will require a `authToken` (_coming soon_) to make an request to the APIs.

## Authentication

Currently, Authentication is under development and will be coming soon.

## Client libraries support

Currently, We only support JavaScript/TypeScript (Node). But, we are looking to support for Go.

---

> # API Reference

Reviewer: **Question:** It looks like these endpoints are all assuming the user point of view, not the team, or installation point of view, is that correct? How extensible will these API's be at the level of team or installation?

Author: [Team APIs](https://deploy-preview-2939--gitpod-kumquat.netlify.app/docs/references/gitpod-public-api#teams-coming-soon) are on the way 🔜

Reviewer: @Siddhant-K-code those look like mostly CRUD operations on top of the teams object, not operations like: 

- List all users in installation (regardless of team)
- List all workspaces (regardless of the user) as an admin

Which would be useful as an administrator of a Gitpod installation.

> **⚠️ Warning:** JS/TS Clients requires atleast [`Node 17`](https://nodejs.org/)

- [Workspaces](#workspaces)
  - [List all workspaces](#list-all-workspaces)
  - [Get a workspace](#get-a-workspace)
  - [Create & start workspace](#create--start-workspace)
  - [Get owner token](#get-owner-token)
  - [Start workspace](#start-workspace)
  - [Stop workspace](#stop-workspace)

> - [Teams (_Coming soon_)](#teams-coming-soon)

Reviewer: https://github.com/gitpod-io/gitpod/pull/14212

  - [Create a team](#create-a-team)
  - List all teams
  - Get a team
  - Update a team
  - Delete a team

---

## Workspaces

### List all workspaces

It lists all the workspaces belongs to the authenticated user.

**Usage**: `listWorkspaces()`

**Request Parameters**:

| Parameter  |                Description                | Type  | Required |
| :--------: | :---------------------------------------: | :---: | :------: |
| `pageSize` | It is maximum number of results we expect | Int32 |  false   |

<br>

> **Response Parameters**:

Reviewer: **Question/suggestion:** What about timestamps for instance when the workspace was last used? This data can be useful for teams who are tracking adoption of the workspaces within an installation / rollout.

|    Parameter    |                                                       Description                                                        |  Type  |
| :-------------: | :----------------------------------------------------------------------------------------------------------------------: | :----: |
| `nextPageToken` |                                                     Next Page Token                                                      | string |

> |  `workspaceId`  |                                                       Workspace Id                                                       | string |

Reviewer: **Thought:** Should this be singular? Should we show the response format when there is multiple items returned?

### Create & start workspace

It creates and start a new workspace to the authenticated user.

**Usage**: `createAndStartWorkspace()`

**Request Parameters**:

|  Parameter   |                Description                |  Type  | Required |
| :----------: | :---------------------------------------: | :----: | :------: |
| `contextUrl` | Context Url - git or prebuild or snapshot | string |   true   |
| `prebuildId` |      Prebuild Id, if prebuild exists      | string |  false   |

<br>

**Response Parameters**:

|   Parameter   |                Description                |  Type  |
| :-----------: | :---------------------------------------: | :----: |
| `workspaceId` | Workspace Id of created/started workspace | string |

### Get owner token

It returns the owner token of a workspace.

**Usage**: `getOwnerToken()`

**Request Parameters**:

|   Parameter   | Description  |  Type  | Required |
| :-----------: | :----------: | :----: | :------: |
| `workspaceId` | Workspace Id | string |   true   |

<br>

**Response Parameters**:

| Parameter | Description |  Type  |
| :-------: | :---------: | :----: |
|  `token`  | Owner Token | string |

> ### Start workspace

Reviewer: **Thought:** Could/should this be `updateWorkspaceStatus` with passed arguments of status, rather than functions for each behaviour?

---
section: gitpod-public-api
title: Gitpod public API
---

<script context="module">
  export const prerender = true;
</script>

# Gitpod public API

> Access Tokens are currently in [Alpha](/docs/help/public-roadmap/release-cycle) · [Send feedback](https://github.com/gitpod-io/gitpod/issues/10798).

The Gitpod Public API grants direct access to Gitpod data for use cases such as integration into other 3rd party applications.

## Authentication

To access the public API, you will need to pass an [Access Token](/docs/configure/user-settings/access-tokens) as a an Authorization request header bearer token in the format `Bearer <your-access-token>`.

## API language clients

### Go

> The following is an example of how to use the Go client library with the Public API for retrieving teams data. Replac the environment variable `PERSONAL_PAT` with your own personal access token, for example: `export PERSONAL_PAT="Bearer <your-personal-pat>"`

Reviewer 3: this pattern is weird. let's make the environment variable we export _just_ the token, and in the code example have the `Bearer ` be prepended in code.

```go
package main

import (
    "context"
    "fmt"
    "net/http"
    "os"

    connect "github.com/bufbuild/connect-go"

    gitpod_experimental_v1 "github.com/gitpod-io/gitpod/components/public-api/go/experimental/v1"
    gitpod_experimental_v1connect "github.com/gitpod-io/gitpod/components/public-api/go/experimental/v1/v1connect"
)

func main() {
    // Define an interceptor to attach credentials onto outgoing requests
    interceptor := connect.UnaryInterceptorFunc(func(next connect.UnaryFunc) connect.UnaryFunc {
        return connect.UnaryFunc(func(ctx context.Context, req connect.AnyRequest) (connect.AnyResponse, error) {
            if req.Spec().IsClient {
                // Send a token with client requests.

>                 req.Header().Set("Authorization", os.Getenv("PERSONAL_PAT"))

Reviewer 3: this is where we'd prepend the `Bearer ` in code to the environment variable

Reviewer 3: ```req.Header().Set("Authorization", "Bearer " + os.Getenv("PERSONAL_PAT"))``` should work
